src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromCmykSimdAvx2.cs

```diff
@@ -0,0 +1,145 @@
+                        "JpegColorConverter.FromGrayscaleVector8 can be used only on architecture having 256 byte floating point SIMD registers!");
+                }
+
+#if SUPPORTS_RUNTIME_INTRINSICS
+                ref Vector256<float> cBase =
+                                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(values.Component0));
+                ref Vector256<float> mBase =
+                                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(values.Component1));
+                ref Vector256<float> yBase =
+                                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(values.Component2));
+                ref Vector256<float> kBase =
+                                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(values.Component3));
+
+                ref Vector256<float> resultBase =
+                    ref Unsafe.As<Vector4, Vector256<float>>(ref MemoryMarshal.GetReference(result));
+
+                // Used for the color conversion
+                var scale = Vector256.Create(1 / maxValue);
+                var one = Vector256.Create(1F);
+
+                // Used for packing
+                ref byte control = ref MemoryMarshal.GetReference(HwIntrinsics.PermuteMaskEvenOdd8x32);
+                Vector256<int> vcontrol = Unsafe.As<byte, Vector256<int>>(ref control);
+
+                int n = result.Length / 8;
+                for (int i = 0; i < n; i++)
+                {
+                    Vector256<float> k = Avx2.PermuteVar8x32(Unsafe.Add(ref kBase, i), vcontrol);
+                    Vector256<float> c = Avx2.PermuteVar8x32(Unsafe.Add(ref cBase, i), vcontrol);
+                    Vector256<float> m = Avx2.PermuteVar8x32(Unsafe.Add(ref mBase, i), vcontrol);
+                    Vector256<float> y = Avx2.PermuteVar8x32(Unsafe.Add(ref yBase, i), vcontrol);
+
+                    k = Avx.Multiply(k, scale);
+
+                    c = Avx.Multiply(Avx.Multiply(c, k), scale);
+                    m = Avx.Multiply(Avx.Multiply(m, k), scale);
+                    y = Avx.Multiply(Avx.Multiply(y, k), scale);
+
+                    Vector256<float> cmLo = Avx.UnpackLow(c, m);
+                    Vector256<float> yoLo = Avx.UnpackLow(y, one);
+                    Vector256<float> cmHi = Avx.UnpackHigh(c, m);
+                    Vector256<float> yoHi = Avx.UnpackHigh(y, one);
+
+                    ref Vector256<float> destination = ref Unsafe.Add(ref resultBase, i * 4);
+
+                    destination = Avx.Shuffle(cmLo, yoLo, 0b01_00_01_00);
+                    Unsafe.Add(ref destination, 1) = Avx.Shuffle(cmLo, yoLo, 0b11_10_11_10);
+                    Unsafe.Add(ref destination, 2) = Avx.Shuffle(cmHi, yoHi, 0b01_00_01_00);
+                    Unsafe.Add(ref destination, 3) = Avx.Shuffle(cmHi, yoHi, 0b11_10_11_10);
+                }
+#else
```

Comment:
Didn't have chance to react, but I'm not entirely happy with the practice established in #1402. I would try putting the different paths into different classes (`FromCmykVector8` VS `FromCmykVectorAvx2`).
`FromCmykVectorAvx2` can derive from `FromCmykVector8` for convenience.

Reply:
Yeah, that was a stopgap. I’d like to revisit

Reply:
If we don't shy away from inheritance here, I would suggest a hierarchy like this:
```csharp
abstract class JpegColorConverter
{
    public abstract bool IsAvailable { get; } // new on this class
}

abstract class BasicJpegColorConverter : JpegColorConverter // new base class for all non-vectorized color converters
{
    public override bool IsAvailable => true;
}

abstract class VectorizedJpegColorConverter : JpegColorConverter // new base class for all vectorized color converters
{
    private readonly int vectorSize;

    protected VectorizedJpegColorConverter(JpegColorSpace colorSpace, int precision, int vectorSize)
        : base(colorSpace, precision)
    {
        this.vectorSize = vectorSize;
    }

    public override void ConvertToRgba(in ComponentValues values, Span<Vector4> result)
    {
        int remainder = result.Length % vectorSize;
        int simdCount = result.Length - remainder;
        if (simdCount > 0)
        {
            ConvertCoreVectorized(values.Slice(0, simdCount), result.Slice(0, simdCount));
        }

        ConvertCore(values.Slice(simdCount, remainder), result.Slice(simdCount, remainder));
    }

    protected abstract void ConvertCoreVectorized(in ComponentValues values, Span<Vector4> result);

    protected abstract void ConvertCore(in ComponentValues values, Span<Vector4> result);
}

abstract class Avx2JpegColorConverter : VectorizedJpegColorConverter // new base class for all AVX2-based converters
{
    protected Avx2JpegColorConverter(JpegColorSpace colorSpace, int precision)
        : base(colorSpace, precision, 8)
    {
    }

    public override bool IsAvailable
    {
        get
        {
#if SUPPORTS_RUNTIME_INTRINSICS
            return Avx2.IsSupported;
#else
            return false;
#endif
        }
    }
}

// another base class for Vector<float>-based converters
```

We could then instantiate all converters initially into a static array and determine which one to use based on `ColorSpace`, `Precision` and `IsAvailable`. This would address your point and maximize code re-use at the cost of some virtual method calls.

Let me know what you think. I didn't want to refactor to much for the initial draft-PR and just add the vectorized logic into the existing code structure.

Reply:
@tkp1n I like the concept, go for it! Hopefully the runtime costs of the virtual dispatches are be still negligible.

Reply:
I refactored as discussed above, PTAL...

Next up: Benchmarks 🚀